### DISCUSSION TITLE
Best practice for handling attachments

### DISCUSSION THREAD

**Discussion by jurica** [zk-org/zk#303] [Help]
Hi,

I'm diving into zk (and zk-nvim) and am looking for a way to handle the following use case:
Often I have attachments (arbitrary "external" files, images, videoclips, log files, emails, etc.) related to a note I want to keep "with it". There is no need for embedding the file into the note, or showing the content/media or anything like that. It's really just about having one or more attachments for a note, kept with it and possible to reference from the note.
My solution for now is just putting the files into a subdirectory "attachments" and have a text reference in the note. But I fear over time this will get messy and hard to keep tidy and organised nicely.

I wonder if anybody else has the same use case and if so how do you handle it?

I'm looking forward for your suggestions, thanks in advance!
Juri

**Comment by stella10x1** [comment]
Hello, as this is a discussion I just share a few thoughts here, but I don't have a absolute solution.

Before zk I used [nb](https://github.com/xwmx/nb). Besides notes (markdown files) nb saved everything inside a git repository. That includes images, videos and all other binary files. Those where indexed just like the notes and where outputted with `nb ls` and could be addressed with the id.  nb could work as a hub for all your files in the nb notebook. E. g. it opens my editor if I want to edit a note or open a pdf viewer if I try to access a pdf file in the notebook.

But I think that solution is incomplete. Also it is no good practice to save big files in git. But it is a good practice to safe the plain text files in git, therefore I recommend doing so with the zk notes as well.

Another issue, which makes working with attachment files harder is, that they aren't accessible via e. g. tags or other metadata. Also you can't link among those files.

One solution would be to create a wrapper note, which links to your actual file. In that notes you could add all your metadata and setup links to other notes. Unfortunately this brings up the problem with e. g. including images in your note, which should be inserted on a html rendered page. It would be desirable to directly add your metadata to the file instead of having two files about that.

An other solution could be [git-annex](https://git-annex.branchable.com/). This is a tool for managing your files in git.  To see how this exactly works please see their description: [how it works](https://git-annex.branchable.com/how_it_works/). In a nutshell it is a *filename and metadata tracker*. It allows you to set tags or other [metadata](https://git-annex.branchable.com/metadata/) to your files. After doing so you could run queries like that to access (git annex tracked) files by metadata:

```bash
git annex find --metadata tag=foo
```

That would be similar to something like `zk list --tag foo` for your notes. Maybe you could write some alias, which outputs the results from `zk list` and `git annex find` at the same time and configure zk to open your PDF viewer if the matched file is a pdf file.

Anyways the git-annex approach does not solve the linking among the files. But you could add an image link to your note.

**Comment by jurica** [comment]
Hi @lingling9000,

thanks for your input and for pointing me towards git-annex!
This seems to be pretty much the git-lfs alternative I was looking for. I store my notes in a git repository that I don't want to pollute with (large) blobs. I should have mentioned this initially...
git-lfs is not an option as the git server (https://github.com/charmbracelet/soft-serve) I'm using doesn't support it. So git-annex might be quite exactly what I'm looking for. I'll take a closer look and give it a try the next days but on first sight this looks very promising.
